pkg/control/sidecarcontrol/sidecarset_control.go

```diff
@@ -143,47 +169,60 @@ func (c *commonControl) IsPodUpdatedConsistent(pod *v1.Pod, ignoreContainers set
 	}
 
 	// check container InpalceUpdate status
-	if err := isSidecarContainerUpdateCompleted(pod, sidecarset.Name, sidecarContainers); err != nil {
-		klog.V(5).Infof(err.Error())
-		return false
-	}
-	return true
+	return isSidecarContainerUpdateCompleted(pod, sidecarset.Name, sidecarContainers)
 }
 
 // k8s only allow modify pod.spec.container[x].image,
 // only when annotations[SidecarSetHashWithoutImageAnnotation] is the same, sidecarSet can upgrade pods
-func (c *commonControl) IsSidecarSetCanUpgrade(pod *v1.Pod) bool {
+func (c *commonControl) IsSidecarSetUpgradable(pod *v1.Pod) bool {
 	sidecarSet := c.GetSidecarset()
-	return GetPodSidecarSetWithoutImageRevision(sidecarSet.Name, pod) == GetSidecarSetWithoutImageRevision(sidecarSet)
+	if GetPodSidecarSetWithoutImageRevision(sidecarSet.Name, pod) != GetSidecarSetWithoutImageRevision(sidecarSet) {
+		return false
+	}
+
+	// cStatus: container.name -> containerStatus.Ready
+	cStatus := map[string]bool{}
+	for _, status := range pod.Status.ContainerStatuses {
+		cStatus[status.Name] = status.Ready
+	}
+	sidecarContainerList := GetSidecarContainersInPod(sidecarSet)
+	for _, sidecar := range sidecarContainerList.List() {
+		// when containerStatus.Ready == true and container non-consistent,
+		// indicates that sidecar container is in the process of being upgraded
+		// wait for the last upgrade to complete before performing this upgrade
+		if cStatus[sidecar] && !c.IsPodUpdatedConsistently(pod, sets.NewString(sidecar)) {
```

Comment:
why we have to wait for the last upgrade to complete? can we upgrade even if the last upgrade is in progress?

Reply:
If we continue to update, it is possible to cause "working" sidecar container to be recreated, hence breaking the service. So waiting for the last upgrade to finish will be a little more business friendly.

> > NeedToInjectInUpdatedPod
> > I also find these two methods are somehow overlapped( IsPodConsistentAndReady and IsPodUpdatedConsistent), maybe we can change IsPodConsistentAndReady to IsPodReady, and make an utility function to implement IsPodConsistentAndReady by using IsPodReady+IsPodUpdatedConsistent

IsPodUpdatedConsistent indicates whether pod.spec and pod.status are consistent after updating the sidecar containers. 
IsPodConsistentAndReady is IsPodUpdatedConsistent + IsPodReady

Reply:
> If we continue to update, it is possible to cause "working" sidecar container to be recreated, hence breaking the service
it is possible to continue updating by directly upgrade the standby sidecar which is undergo a previous upgrading.